lightning/src/routing/router.rs

```diff
@@ -1643,96 +1643,57 @@ where L::Target: Logger {
 		return Err(LightningError{err: "Failed to find a sufficient route to the given destination".to_owned(), action: ErrorAction::IgnoreError});
 	}
 
-	// Sort by total fees and take the best paths.
-	payment_paths.sort_unstable_by_key(|path| path.get_total_fee_paid_msat());
-	if payment_paths.len() > 50 {
-		payment_paths.truncate(50);
-	}
+	// Step (6).
+	let mut selected_route = payment_paths;
 
-	// Draw multiple sufficient routes by randomly combining the selected paths.
-	let mut drawn_routes = Vec::new();
-	let mut prng = ChaCha20::new(random_seed_bytes, &[0u8; 12]);
-	let mut random_index_bytes = [0u8; ::core::mem::size_of::<usize>()];
+	debug_assert_eq!(selected_route.iter().map(|p| p.get_value_msat()).sum::<u64>(), already_collected_value_msat);
+	let mut overpaid_value_msat = already_collected_value_msat - final_value_msat;
 
-	let num_permutations = payment_paths.len();
-	for _ in 0..num_permutations {
-		let mut cur_route = Vec::<PaymentPath>::new();
-		let mut aggregate_route_value_msat = 0;
+	// First, sort by the cost-per-value of the path, selecting only the paths which
+	// contribute the most per cost.
+	// We sort backwards as we will remove from the front in `retain`, next.
+	selected_route.sort_unstable_by(|a, b|
+		(((b.get_cost_msat() as u128) << 64) / (b.get_value_msat() as u128))
+			.cmp(&(((a.get_cost_msat() as u128) << 64) / (a.get_value_msat() as u128)))
+	);
 
-		// Step (6).
-		// Do a Fisher-Yates shuffle to create a random permutation of the payment paths
-		for cur_index in (1..payment_paths.len()).rev() {
-			prng.process_in_place(&mut random_index_bytes);
-			let random_index = usize::from_be_bytes(random_index_bytes).wrapping_rem(cur_index+1);
-			payment_paths.swap(cur_index, random_index);
+	// We should make sure that at least 1 path left.
+	let mut paths_left = selected_route.len();
+	selected_route.retain(|path| {
+		if paths_left == 1 {
+			return true
+		}
+		let mut keep = true;
+		let path_value_msat = path.get_value_msat();
+		if path_value_msat <= overpaid_value_msat {
+			keep = false;
+			overpaid_value_msat -= path_value_msat;
+			paths_left -= 1;
 		}
+		keep
+	});
+	assert!(selected_route.len() > 0);
```

Comment:
Should this be a `debug_assert`? If we hit this something went terribly wrong, but I'm not sure we're helping the user if we're crashing on top of it?